src/core/models/inputs/utxo_input.c

```diff
@@ -1,108 +1,290 @@
 // Copyright 2021 IOTA Stiftung
 // SPDX-License-Identifier: Apache-2.0
 
-#include <inttypes.h>
 #include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
 
 #include "core/models/inputs/utxo_input.h"
+#include "core/models/outputs/outputs.h"
+#include "utlist.h"
 
-#define UTXO_INPUT_MIN_INDEX 0
-#define UTXO_INPUT_MAX_INDEX 126
+utxo_inputs_list_t *utxo_inputs_new() { return NULL; }
 
-int utxo_inputs_add_with_key(utxo_input_ht **inputs, byte_t const tx_id[], uint16_t index, byte_t const pub[],
-                             byte_t const priv[]) {
-  if (index > UTXO_INPUT_MAX_INDEX) {
+void utxo_inputs_free(utxo_inputs_list_t *inputs) {
+  if (inputs) {
+    utxo_inputs_list_t *elm, *tmp;
+    LL_FOREACH_SAFE(inputs, elm, tmp) {
+      free(elm->input);
+      LL_DELETE(inputs, elm);
+      free(elm);
+    }
+  }
+}
+
+int utxo_inputs_add(utxo_inputs_list_t **inputs, uint8_t type, byte_t id[], uint16_t index) {
+  if (id == NULL) {
+    printf("[%s:%d] invalid transaction id\n", __func__, __LINE__);
+    return -1;
+  }
+
+  if (index <= UTXO_INPUT_MIN_INDEX || index >= UTXO_OUTPUT_MAX_COUNT) {
```

Comment:
tx output index must be `0 ≤ x < 127`, index 0 should be accepted by this method. 
since `index` is an uint16_t:
```suggestion
  if (index >= UTXO_OUTPUT_MAX_COUNT) {
```

Reply:
Done.